### ISSUE TITLE
xen-hvm  GNU/Linux instances: type=ioemu into vif definitions breaks instances network

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#303]
Originally reported of Google Code with ID 247.

```
What software version are you running? Please provide the output of "gnt-
cluster --version" and "gnt-cluster version".

<b>What distribution are you using?</b>

Debian GNU/Linux 6.0.5 (squeeze) x86_64

<b>What steps will reproduce the problem?</b>
1. Create an xen-hvm instance with an qemu emulated vNIC (default rtl8139) and attach it a debian-6.x netinst iso:<code>
root@gnt-node1:~# gnt-instance add --hypervisor-parameters xen-hvm:boot_order=dc,cdrom_image_path=/usr/share/ganeti/os/iso/debian-6.0.4-amd64-i386-netinst.iso --no-start --no-install --disk-template drbd --os-type debootstrap+default --disk 0:size=5G --backend-parameters memory=512 --node gnt-node1.example.com:gnt-node2.example.com  gnt-instance-2.example.com

root@gnt-node1:~# gnt-instance start gnt-instance-2.example.com

root@gnt-node1:/etc# gnt-instance info gnt-instance-2.example.com
Instance name: gnt-instance-2.example.com
UUID: e89e09c1-a570-4518-8023-3a11a2bbbd1a
Serial number: 1
Creation time: 2012-06-23 20:26:19
Modification time: 2012-06-23 20:26:19
State: configured to be down, actual state is down
  Nodes:
    - primary: gnt-node1.example.com
    - secondaries: gnt-node2.example.com
  Operating system: debootstrap+default
  Allocated network port: 11008
  Hypervisor: xen-hvm
    - console connection: vnc to gnt-node1.example.com:11008 (display 5108)
    - acpi: default (True)
    - blockdev_prefix: default (xvd)
    - boot_order: dc
    - cdrom_image_path: /usr/share/ganeti/os/iso/debian-6.0.4-amd64-i386-netinst.iso
    - device_model: default (/usr/lib/xen-4.0/bin/qemu-dm)
    - disk_type: default (paravirtual)
    - kernel_path: default (/usr/lib/xen-4.0/boot/hvmloader)
    - nic_type: default (rtl8139)
    - pae: default (True)
    - reboot_behavior: default (reboot)
    - use_localtime: default (False)
    - vnc_bind_address: default (0.0.0.0)
    - vnc_password_file: default (/etc/ganeti/vnc-cluster-password)
  Hardware:
    - VCPUs: 1
    - memory: 512MiB
    - NICs:
      - nic/0: MAC: aa:00:00:db:c6:a4, IP: None, mode: bridged, link: xen-br30
  Disk template: drbd
  Disks:
    - disk/0: drbd8, size 5.0G
      access mode:  rw
      nodeA:        gnt-node1.example.com, minor=1
      nodeB:        gnt-node2.example.com, minor=1
      port:         11012
      auth key:     39db28d7f531978115ac2b4f09e0f6d57a8cc9a2
      on primary:   /dev/drbd1 (147:1) in sync, status ok
      on secondary: /dev/drbd1 (147:1) in sync, status ok
      child devices:
        - child 0: lvm, size 5.0G
          logical_id:   xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_data
          on primary:   /dev/xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_data (254:2)
          on secondary: /dev/xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_data (254:2)
        - child 1: lvm, size 128M
          logical_id:   xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_meta
          on primary:   /dev/xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_meta (254:3)
          on secondary: /dev/xenvg/4d2bfe4a-04f1-40f7-bb6a-55ca976515c3.disk0_meta (254:3)

</code>
2. Start the instance from CD, connect to it with an VNC Client and proceed with an x86_64 graphical installation 
3. Check that instance is unable to acquire ip address from dhcp:
  - "ip link show" should list an eth0 paravirtualized interfaces but with a mac address full of zeroes 
  - the following error message is displayed:  "rtnetlink answers: cannot assign requested address" 
  - xen pvhvm network driver is used (aka xen_netfront) even if we specified an qemu emulated nic. 


What is the expected output? 
Instance should be able to get ip settings either manually or from dhcp

What do you see instead?
I'm unable to assign ipv4 network settings to the xen-hvm linux instance

<b>Please provide any additional information below.</b>

- This issue is a follow up to this discussion: https://groups.google.com/forum/?fromgroups#!topic/ganeti/BB26ogOagMA

- There's a warning on xen official wiki about this specific type=ioemu issue: http://wiki.xen.org/wiki/Xen_Linux_PV_on_HVM_drivers#Xen_PVHVM_drivers_configuration_example
And this is confirmed by this ubuntu bug https://bugs.launchpad.net/ubuntu/+bug/949028 

- To workaround this issue, one just have to "patch" /usr/share/pyshared/ganeti/hypervisor/hv_xen.py and destroy/recreate the xen-hvm instance :<code>
root@gnt-node1:/usr/share/pyshared/ganeti/hypervisor# diff -u hv_xen.py.orig hv_xen.py
--- hv_xen.py.orig	2012-06-24 17:22:02.000000000 +0200
+++ hv_xen.py	2012-06-24 17:39:29.000000000 +0200
@@ -650,7 +650,7 @@
     elif nic_type == constants.HT_NIC_PARAVIRTUAL:
       nic_type_str = ", type=paravirtualized"
     else:
-      nic_type_str = ", model=%s, type=ioemu" % nic_type
+      nic_type_str = ", model=%s" % nic_type
     for nic in instance.nics:
       nic_str = "mac=%s%s" % (nic.mac, nic_type_str)
       ip = getattr(nic, "ip", None)

</code> 

 - Was able to confirm the same network issue with an CentOS x86_64  and Ubuntu 12.04 instances 

-  To sum up the various test done on my 3-nodes Debian squeeze x86_64 cluster:
  - debian netinst x86_64 + nic_type={e1000,ne2k_pci,rtl8139} => NOK
  - debian netinst x86_64 + nic_type=paravirtual => OK
  - debian netinst x86 + nic_type={e1000,ne2k_pci,rtl8139} => OK
  - debian netinst x86 + nic_type=paravirtual => NOK . This one may be normal as i don't know if paravirtualized disk and network drivers are available for x86 instances

HTH,
Thanks for Ganeti !  

```
Originally added on 2012-06-27 22:15:14 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ooops, just forgot to add that ganeti feature request #199 (http://code.google.com/p/ganeti/issues/detail?id=199 ) is may be related to this bug report as its mentionned on xen official wiki 

Thanks ! 
```
Originally added on 2012-06-27 22:19:17 +0000 UTC.

Mentioned Issue with GoogleCode ID 199 has been migrated to GitHub with ID #255.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Thanks for the report, this indeed looks like it needs fixing!

Since this needs to be done on a per-xen version, we can't simply remove the type, but we should look at the Xen version and do the appropriate thing.

Also interesting thread: http://lists.xen.org/archives/html/xen-devel/2012-06/msg01595.html
```
Originally added on 2012-07-19 15:45:46 +0000 UTC.

Added Labels: Type-Defect Hypervisor-XenHVM
Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hello,
i can reproduce the same problem trying to run a Centos 6.2 HVM guest under ganeti+xen. 
There is no NIC in the guest if the vif definition includes type=ioemu.
Manually removing "type=ioemu" from the xen config file of the guest solves the problem.
```
Originally added on 2012-11-25 11:00:17 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
No updates on this issue?
```
Originally added on 2013-02-28 15:28:01 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Not yet, sorry.
I'll try to put it among the ones in the 2.7 beta series, and see if I can find a bit of time to work on it, but it might slip to 2.8.

Of course if there was some volunteer that could update the patch to take the Xen version into consideration and do the right thing on every version we could push it in now.

Thanks,
Guido
```
Originally added on 2013-02-28 18:09:14 +0000 UTC.

Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-04-10 14:59:43 +0000 UTC.

Added Labels: Priority-Medium

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-04-26 08:27:00 +0000 UTC.

Added Labels: Priority-High
Removed Labels: Priority-Medium

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-04-29 13:29:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit c54c859b3e48e9877f1043c8e731d4ebd215b3c8
Author: Helga Velroyen <helgav@google.com>
Date:   Mon Apr 29 18:43:39 2013 +0200

    man gnt-instance: new hypervisor parameter 'vif_type'
    
    This patch extends the man page of the 'gnt-instance' command
    by describing the newly implemented hypervisor parameter
    'vif_type'.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Guido Trotter <ultrotter@google.com>

commit 0544084580b272851247afb088e2714104f28314
Author: Helga Velroyen <helgav@google.com>
Date:   Mon Apr 29 17:04:29 2013 +0200

    gnt-instance: new hypervisor parameter 'vif_type'
    
    This patch fixes issue 247. It empowers the user to specify
    explicitly the vif type of a nic configuration of a Xen HVM
    instance. This includes the option of setting it to '' on the
    commandline, causing Ganeti to leave out the 'type' parameter
    in the vif configuration completely.
    
    Signed-off-by: Helga Velroyen <helgav@google.com>
    Reviewed-by: Guido Trotter <ultrotter@google.com>
```
Originally added on 2013-04-30 07:46:13 +0000 UTC.

Mentioned Issue with GoogleCode ID 247 has been migrated to GitHub with ID #303.
Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
